File: backend_go/models/payment_invoice.go
package models

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

type InvoiceStatus string

const (
	InvoiceStatusPending         InvoiceStatus = "pending"
	InvoiceStatusCompleted       InvoiceStatus = "completed"
	InvoiceStatusFailed          InvoiceStatus = "failed"
	InvoiceStatusAwaitingCheck   InvoiceStatus = "awaiting_check"
	InvoiceStatusManuallyConfirmed InvoiceStatus = "manually_confirmed"
)

// PaymentInvoice represents a record of a payment attempt.
type PaymentInvoice struct {
	ID        uint `gorm:"primaryKey"`
	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt gorm.DeletedAt `gorm:"index" swaggertype:"string" example:"2024-01-01T00:00:00Z"`

	BotUserID uint    `gorm:"index"`
	BotUser   BotUser `gorm:"foreignKey:BotUserID"`

	OriginalAmount      float64
	Amount              float64
	Status              InvoiceStatus   `gorm:"index"`
	Gateway             string          `gorm:"index"`
	GatewayInvoiceID    string          `gorm:"index"`
	OrderID             string          `gorm:"uniqueIndex"` // Our internal unique ID for the transaction
	PayURL              *string         `json:"pay_url"`
	PaymentDetails      json.RawMessage `gorm:"type:jsonb" json:"payment_details" swaggertype:"object"` // Store raw JSON from gateway
	BotMessageID        *int64
	WasNotificationSent bool `gorm:"default:false"`
}

File: backend_rust/src/presentation/admin/dtos/payment_invoice.rs
use rust_decimal::prelude::ToPrimitive;
use shared_dtos::invoice::PaymentInvoiceAdminResponse;

use crate::models::payment_invoice::PaymentInvoiceRow;

impl From<PaymentInvoiceRow> for PaymentInvoiceAdminResponse {
    fn from(r: PaymentInvoiceRow) -> Self {
        PaymentInvoiceAdminResponse {
            id: r.id,
            customer_id: r.customer_id,
            original_amount: r.original_amount.to_f64().unwrap_or_default(),
            amount: r.amount.to_f64().unwrap_or_default(),
            status: r.status,
            created_at: r.created_at,
            updated_at: r.updated_at,
            expires_at: r.expires_at,
            gateway: r.gateway,
            gateway_invoice_id: r.gateway_invoice_id,
        }
    }
}
